2139 add maxpy and mdot support

prudhomm opened the pull request:
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/feelpp/feelpp/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes?
- [x] Have you successfully run the Feel++ testsuite with your changes locally?
- [x] Have you written Doxygen comments in your contribution ?

-----

- closes #1912
- closes #1950 
- closes #2038
- closes #2041
- closes #2125  
- closes #2139

thomas-saigre committed:
tiny fixes in test

thomas-saigre committed:
Merge branch '1912-implement-nirb-tools' of github.com:feelpp/feelpp into 1912-implement-nirb-tools
[ci skip]

thomas-saigre committed:
start work on greedy with nirb approximation
[ci skip]

thomas-saigre committed:
up greedy code #2038
the architecture is present but still bugs in it
[ci skip]

thomas-saigre committed:
up greedy algorithm #2038
[ci skip]

thomas-saigre committed:
add function getRectificationMatrix #1912
/cc @alielarif
[skip feelpp, skip tests, skip toolboxes, skip mor]

thomas-saigre committed:
up forgotten modif in 92674b0b74f59749ccc6ac43e91dc550db70da02
#1912
[ci skip]

thomas-saigre committed:
refactor nirb tests #2038
[ci skip]

thomas-saigre committed:
refactor nirbOffline script and tests
now scripts Offline.py and Online.py are installed and can be loaded.
The function implemented are used in test_nirb

for now only offline has been refactored

[ci skip]

alielarif committed:
up wrapping in sampling

alielarif committed:
get distinct sampling for test and train

alielarif committed:
rename file skip feelpp skip tests skip toolboxes skip mor

alielarif committed:
clean up file

alielarif committed:
Merge branch '2014-add-multiblock-2d-and-3d-use-case-generator' into 2041-test-nirb-convergence

thomas-saigre committed:
fix greedy+rectificatio tests #2038
[ci skip]

thomas-saigre committed:
Fix tests : add online tests #2038 #1912
[ci skip]

thomas-saigre committed:
refactor online #1912 #2038
[ci skip]

alielarif committed:
remove rectification param from offline. test nirb convergence #2041 implement nirb tools in feel++ #1912

alielarif committed:
test nirb perf in respect of rectif parameter.

alielarif committed:
remove rectif param from offline

alielarif committed:
up compare list of errors

alielarif committed:
up file

alielarif committed:
up plot lsit of errors

alielarif committed:
create dir by one proc.

alielarif committed:
add thermal bridge test. skip feelpp skip tests skip toolboxes skip mor test nirb convergence #2041

alielarif committed:
Merge branch '2014-add-multiblock-2d-and-3d-use-case-generator' into 2041-test-nirb-convergence

thomas-saigre committed:
add function greedy step #2038

thomas-saigre committed:
tiny fixes
[ci skip]

thomas-saigre committed:
add class method to export basis function
[ci skip]

alielarif committed:
up printing infos. skip feelpp skip tests skip toolboxes skip mor

thomas-saigre committed:
add functions to export basis and nirb solutions for different size of basis

thomas-saigre committed:
Merge remote-tracking branch 'origin/1912-implement-nirb-tools' into 2038-nirb-greedy

thomas-saigre committed:
fix typos and minor errors
[ci skip]

thomas-saigre committed:
refactor createInterpolator function
[ci skip]

thomas-saigre committed:
refactor : use np.linalg.solve instead of a for loop
[ci skip]

alielarif committed:
Merge remote-tracking branch 'origin/develop' into 2041-test-nirb-convergence

thomas-saigre committed:
Merge remote-tracking branch 'origin/develop' into 2038-nirb-greedy

alielarif committed:
add time parameters

alielarif committed:
add option for error computation

alielarif committed:
reduce variables skip feelpp skip tests skip toolboxes skip mor

alielarif committed:
load online nirb parameters from offlineFileInfos. skip feelpp skip tests skip toolboxes skip mor

alielarif committed:
load online nirb parameters from offlineFileInfos. skip feelpp skip tests skip toolboxes skip mor

alielarif committed:
clean up

alielarif committed:
show interp error. skip feelpp skip tests skip toolboxes skip mor

alielarif committed:
rename files and remove hardcoded. skip feelpp skip tests skip toolboxes skip mor

alielarif committed:
Merge branch '2041-test-nirb-convergence' of github.com:feelpp/feelpp into 2041-test-nirb-convergence

alielarif committed:
remove dowloded files from workspace. skip feelpp skip tests skip toolboxes skip mor

thomas-saigre committed:
Merge remote-tracking branch 'origin/develop' into 2038-nirb-greedy

thomas-saigre committed:
fix tests after merging the branch
[ci skip]

thomas-saigre committed:
Merge remote-tracking branch 'origin/develop' into 2041-test-nirb-convergence

thomas-saigre committed:
fix nirb executions in notebook

[ci skip]

thomas-saigre committed:
Merge branch '2041-test-nirb-convergence' into 2038-nirb-greedy

thomas-saigre committed:
up fixes in greedy function
[ci skip]

thomas-saigre committed:
add __init__.py to load nirb library
[ci skip]

thomas-saigre committed:
refactor nirb offline script

[ci skip]

thomas-saigre committed:
Merge branch '2038-nirb-greedy' of github.com:feelpp/feelpp into 2038-nirb-greedy
[ci skip]

thomas-saigre committed:
rename function initProblem -> computeSnapshots
#2125

[ci skip]

thomas-saigre committed:
refactor PODReducedBasis, add some comments in code #2125

the attribute self.ReducedBasis is updated inplace, as well as self.N (the size of the basis)

thomas-saigre committed:
refactor : use nirb_online in check convergence function
#2125

still issues with rectification

thomas-saigre committed:
fix computation of rectification matrix
#2125

thomas-saigre committed:
up tiny fixes  in files

thomas-saigre committed:
refactor nirb_perf :
- computeError now calls computeErrorSampling
- remove duplicate of code in computeErrorSampling and checkConvergence
- remove getNirbProjection that was a code duplicate
- add function fineProjection in nirbOnline to compute the projection of fine snapshot
in the reduced basis

thomas-saigre committed:
fix greedy execution
(the function runs well, but no tests on results has been made yet)

 #2125

thomas-saigre committed:
fix nirbOnline code,
and export offline info in order to read it in online stage

thomas-saigre committed:
fix notebooks for both offline and online

thomas-saigre committed:
fix nirb tests

thomas-saigre committed:
Merge remote-tracking branch 'origin/develop' into 2125-fix-nirb-execution
[ci skip]

thomas-saigre committed:
fix FEELPP_ENABLE_MOR
/cc @vincentchabannes

[ci skip]

prudhomm committed:
skip feelpp skip tests skip toolboxes skip mor

thomas-saigre committed:
fix tests #2125
[skip feelpp, skip mor, skip toolboxes, skip tests]

thomas-saigre committed:
refactor and fix tests #2125
and add tic and toc to measure snapshots computation time

[skip feelpp, skip toolboxes, skip tests, skip mor]

thomas-saigre committed:
add tic / toc in offline and online function
[ci skip]

thomas-saigre committed:
other tiny fixes in notebook
[ci skip]

thomas-saigre committed:
fix creaton of directory in parallel :
only master creates it
[ci skip]

thomas-saigre committed:
compute H1 norm too
[ci skip]

thomas-saigre committed:
up code to post process error computed in order to create csv file
[ci skip]

thomas-saigre committed:
some tiny fixes :
- export Xi_train
- compute rectification matrix with regularization parameter
- fix getRectification

[ci skip]

thomas-saigre committed:
add function to compute and save reduced basis functions #1950
[ci skip]

thomas-saigre committed:
refactor : move common function into ToolboxModel class
- use function getFineSolution and getCoarseSolution
- generate operators, fine and coarse, h1 and l2
#2125
[ci skip]

thomas-saigre committed:
fix execution and tests :
- systematically call to reduceBasisFunctions
- fix typos
- set coarse function space XH
[ci skip]

thomas-saigre committed:
up :
- add function to compute alphaNH
- check that vecotr saved are correct
- compute alpha of the right size
[ci skip]

thomas-saigre committed:
up test to check that data are corectly saved
[ci skip]

thomas-saigre committed:
fix load of the coefficients matrices for rectification
#2125
[ci skip]

thomas-saigre committed:
cancel wrong improvments
[ci skip]

thomas-saigre committed:
use PETSc fuctions instead of loops #1950
work still need to be done for computation pf alpha_i #2139
/cc @prudhomm
[ci skip]

thomas-saigre committed:
up mDot and maxpy functions #2139
/cc @prudhomm
[ci skip]

prudhomm committed:
Merge branch 'develop' into 2139-add-maxpy-and-mdot-support

prudhomm committed:
implement multiple dot products and add a linear combination of vectors interface

- update python interface
fixes #2139
/cc @thomas-saigre

prudhomm committed:
make the interface more consistent

renamed newElements to newVectors since the data structure is a set of vectors that need to be cast back to element

updated python interface

/cc @thomas-saigre

prudhomm commented:
@thomas-saigre the build is broken here https://buildkite.com/feelpp/feelpp/builds/4977#0189550a-da89-423e-9cd7-5fef584b3348/413-1610 

this is a compilation issue, could you have a look ? I have no problem in the `develop` branch with this code

thomas-saigre committed:
fix compilation #2139
use Feel namespace to avoid ambiguity
/cc @prudhomm
[skip feelpp, skip tests, skip toolboxes]

thomas-saigre committed:
up functions to test mDot
#2139
/cc @prudhomm

[ci skip]

thomas-saigre commented:
@prudhomm It also seems that the wrapping has been broken : https://buildkite.com/feelpp/feelpp/builds/4980#01895eda-c34a-41c0-b040-e791c442f410/695-1122

prudhomm committed:
include eigen.h to convert Eigen data structures to python data structure

/cc @thomas-saigre
skip feelpp skip tests skip toolboxes skip mor
#2139

thomas-saigre committed:
up test_alg #2139
[ci skip]

thomas-saigre committed:
up test discr tests for mDot #2139
add header to stl for conversion

test not passing yet
/cc @prudhomm

[ci skip]

thomas-saigre committed:
use vectorized functions instead of loops #2139
[ci skip]

thomas-saigre committed:
add test for the addfunction takin float as parameter #2139
/cc @prudhomm
[ci skip]

thomas-saigre committed:
attemp to fix bug induced by the adding of pybind/stl.h

convert by hand the feelpp._core.WorldsComm in a List[feelpp._core.WroldComm]

[skip feelpp, skip toolboxes, skip mor, skip tests]

/cc @prudhomm

thomas-saigre committed:
rename add(float, vector) -> addVector(float, vector)
because the wrapping of add(List[float], List[vector]) messed it up
/cc @prudhomm
[skip feelpp, skip tests, skip mor, skip toolboxes]

thomas-saigre commented on python/pyfeelpp/feelpp/__init__.py.in:
```diff
@@ -175,10 +175,11 @@ def functionSpace( mesh, space="Pch", order=1, worldscomm=None):
     """
     if worldscomm is None:
         worldscomm = Environment.worldsComm(1)
+    worldscomm_list = [worldcomm for worldcomm in worldscomm]
```
@prudhomm With the adding of the header to `pybind11/stl.h`, the constructor for `spaces[key]` was no longer waiting a `feelpp._core.WorldsComm` but a `List[feelpp._core.WorldComm]`.
I found this workaround by converting it by hand to a Python list, but there may be a better way to do it...

#2139

thomas-saigre commented on python/pyfeelpp/feelpp/alg.cpp:
```diff
@@ -276,7 +276,7 @@ PYBIND11_MODULE(_alg, m )
         ( std::shared_ptr<Vector<double>> const& , std::shared_ptr<MatrixSparse<double>> const&  )
         >( &Vector<double>::addVector ), "add the product of a MatrixSparse and a vector to this vector" )
-        .def( "add", [](Vector<double> &v, double a, Vector<double> const& w){ v.add(a, w); }, "add a scalar*vector to this vector" )
+        .def( "addVector", [](Vector<double> &v, double a, Vector<double> const& w){ v.add(a, w); }, "add a scalar*vector to this vector" )
```
@prudhomm With the wrapping `add(List[float] alphas, List[vector] vs)` computing $u += \sum_i \alpha_i vs[i]$, the initial `add(float alpha, vector v)` computing $u += \alpha u$ was no longer included in the wrapped class.
A workaround I found is to rename this method to `addVector` so it can be wrapped correctly. Maybe there is another way to do it...

#2139

thomas-saigre committed:
remove useless print
[ci skip]

thomas-saigre committed:
fixes for nirb :
- fix offline execution and refactor code with removing some loops
- fix onlien execution : used the right number of basis functions

#2139

[skip feelpp, skip toolboxes, skip mor, skip tests]

thomas-saigre committed:
fix time measurment

[skip feelpp, skip mor, skip tests, skip toolboxes]

prudhomm committed:
Merge branch 'develop' into 2139-add-maxpy-and-mdot-support

prudhomm commented:
@thomas-saigre I sync'ed with develop. there is still work to do in mor

thomas-saigre committed:
tiny modifications for merge

[ci skip]

thomas-saigre committed:
Merge branch '2139-add-maxpy-and-mdot-support' of github.com:feelpp/feelpp into 2139-add-maxpy-and-mdot-support
[ci skip]

thomas-saigre committed:
fix s.sampling -> s.sample

[skip feelpp, skip tests]

thomas-saigre committed:
Merge remote-tracking branch 'origin/develop' into 2139-add-maxpy-and-mdot-support

prudhomm committed:
Merge branch 'develop' into 2139-add-maxpy-and-mdot-support

thomas-saigre committed:
Merge branch '2139-add-maxpy-and-mdot-support' of github.com:feelpp/feelpp into 2139-add-maxpy-and-mdot-support
